src/bin/stats.rs

```diff
@@ -0,0 +1,104 @@
+use std::collections::HashMap;
+use std::fs;
+
+use color_eyre::eyre::Result;
+use miri_the_world::*;
+
+fn main() -> Result<()> {
+    if std::env::var("RUST_BACKTRACE").is_err() {
+        std::env::set_var("RUST_BACKTRACE", "1");
+    }
+    if std::env::var("RUST_LOG").is_err() {
+        std::env::set_var("RUST_LOG", "info");
+    }
+    env_logger::init();
+    color_eyre::install()?;
+
+    let mut crates = HashMap::new();
+    for line in fs::read_to_string("crates.json")?.lines() {
+        let krate: Crate = serde_json::from_str(&line)?;
+        crates.insert(krate.name.clone(), krate);
+    }
+
+    let mut times = vec![];
+    let mut total_time = 0;
+    let mut states: HashMap<_, usize> = HashMap::new();
+    let mut errored = 0;
+    let mut ub = 0;
+    let mut known = 0;
+    for krate in crates.values() {
+        total_time += krate.time;
+        match &krate.status {
+            Status::Unknown => continue,
+            Status::Passing => {
+                known += 1;
+            }
+            Status::Error(_) => {
+                errored += 1;
+                known += 1;
+                continue;
+            }
+            Status::UB { cause: causes, .. } => {
+                for cause in causes {
+                    *states.entry(cause.kind.clone()).or_default() += 1;
+                }
+                ub += 1;
+                known += 1;
+            }
+        }
+        let mut time = krate.time as usize / 60;
+        let rm = krate.time as usize % 60;
+        if rm != 0 {
+            time += 1;
+        }
+        if times.len() <= time {
+            times.resize(time + 1, 0);
```

Comment:
I tried this out on 1,000 crates with a 10,000 crate file. Since the `time` field defaults to `u64::max_value()`, this line fails to allocate, and aborts. I think the `time` field should be an `Option`, or it should be rolled into the enum to prevent this.

Reply:
It defaults to `0`, so this will happen if there are humongous values in the field already (which there aren't because you don't have it yet). Not sure what happened here, do you have more info?

Reply:
The default to `u64::max` is here: https://github.com/saethlin/miri-tools/pull/2/files#diff-42cb6807ad74b3e201c5a7ca98b911c5fa08380e942be6e4ac5807f8377f87fcR84

I think this was previously caused by the discrepancy between the length of `crates.json` and the value of `--crates`.

Overall, I'm less and less excited about holding on to a crate list. The state we actually want is in the build logs. The crate list is really nice in terms of startup time, but the download rankings rot very quickly (`vsdb` has plummeted since they stopped whatever they were doing or were blocked), and since runs do take some time from run to run it seems advisable to have the latest version of all crates, which is _incredibly slow_ to get from the crates.io API because you have to do a crate at a time.